src/linear_algebra/bilinear_form.lean

```diff
@@ -1016,3 +1031,331 @@ begin
+  { exact submodule.mem_span_singleton.2 ⟨(B x y * (B x x)⁻¹), rfl⟩ },
+  { intros z hz,
+    rcases submodule.mem_span_singleton.1 hz with ⟨μ, rfl⟩,
+    simp [is_ortho, mul_assoc, inv_mul_cancel hx, hB x] },
+  { simp }
+end
+
+/-- Given a bilinear form `B` and some `x` such that `B x x ≠ 0`, the span of the singleton of `x`
+  is complement to its orthogonal complement. -/
+lemma is_compl_singleton {B : bilin_form K V}
+  (hB : sym_bilin_form.is_sym B) {x : V} (hx : B x x ≠ 0) :
+  is_compl (submodule.span K ({x} : set V))
+    (B.orthogonal (submodule.span K ({x} : set V))) :=
+{ inf_le_bot := eq_bot_iff.1 $ span_inf_ortho_eq_bot hx,
+  top_le_sup := eq_top_iff.1 $ span_sup_ortho_eq_top hB hx }
+
+/-- The natural isomorphism between a singleton and the quotient by its orthogonal complement. -/
+noncomputable def quotient_equiv_of_ortho_singleton
+{B : bilin_form K V} (hB : sym_bilin_form.is_sym B) {x : V} (hx : B x x ≠ 0) :=
+submodule.quotient_equiv_of_is_compl _ _ (is_compl_singleton hB hx)
+
+/-- The natural isomorphism from the product between a singleton and its orthogonal component
+  and the whole space. -/
+noncomputable def prod_equiv_of_ortho_singleton
+  {B : bilin_form K V} (hB : sym_bilin_form.is_sym B) {x : V} (hx : B x x ≠ 0) :=
+  submodule.prod_equiv_of_is_compl _ _ (is_compl_singleton hB hx)
+
+lemma restrict_ortho_singleton_nondegenerate (B : bilin_form K V) (hB₁ : nondegenerate B)
+  (hB₂ : sym_bilin_form.is_sym B) {x : V} (hx : B x x ≠ 0) :
+  nondegenerate $ B.restrict $ B.orthogonal (submodule.span K ({x} : set V)) :=
+begin
+  refine λ m hm, submodule.coe_eq_zero.1 (hB₁ m.1 (λ n, _)),
+  have : n ∈ submodule.span K ({x} : set V) ⊔
+    B.orthogonal (submodule.span K ({x} : set V)) :=
+    (span_sup_ortho_eq_top hB₂ hx).symm ▸ submodule.mem_top,
+  rcases submodule.mem_sup.1 this with ⟨y, hy, z, hz, rfl⟩,
+  specialize hm ⟨z, hz⟩,
+  rw [restrict_def, subtype.val_eq_coe] at hm,
+  erw [add_right, show B m.1 y = 0, by exact m.2 y hy, hm, add_zero]
+end
+
+/- Let V be a finite dimensional vector space over the field K with the
+  nondegenerate bilinear form B. Then for all m ∈ M, f_m : M → R₁ : n ↦ B m n is
+  a linear functional in the dual space.
+
+  Furthermore, the map, φ : M → M* : m ↦ f_m is an isomorphism. -/
+
+/-- Given a bilinear form `B`, `to_dual_aux` maps elements `m` of the module `M`
+  to the functional `λ x, B m x` in the dual space. -/
+def to_dual_aux (B : bilin_form R₃ M₃) (m : M₃) : module.dual R₃ M₃ :=
+{ to_fun := λ n, B m n,
+  map_add' := add_right m,
+  map_smul' := λ _ _, by simp only [algebra.id.smul_eq_mul, smul_right] }
+
+@[simp] lemma to_dual_aux_def {B : bilin_form R₃ M₃} {m n : M₃} :
+  B.to_dual_aux m n = B m n := rfl
+
+/-- Given a bilinear form `B` on the modual `M`, `to_dual' B` is the linear map
+  from `M` to its dual such that `to_dual B m` is the functional `λ x, B m x`. -/
+def to_dual' (B : bilin_form R₃ M₃) : M₃ →ₗ[R₃] module.dual R₃ M₃ :=
```

Comment:
As discussed on discord, `to_dual` is defeq to `to_lin` - so should probably be defined in terms of it.

Reply:
Can you summarize?  Why does `to_dual'` need to exist?  Does it work simply to refer to `B.to_lin` where needed, eg
```lean
noncomputable def to_dual (B : bilin_form K V) (hB : B.nondegenerate) :
  V ≃ₗ[K] module.dual K V :=
{ map_smul' := B.to_lin.map_smul',
```
etc?

Reply:
My comment is outdated, it referred to the lines below that have since been replaced. The implementation below of `B.to_lin` is exactly what I was asking for.

Reply:
I will use the method @hrmacbeth said in her last comment so we won't need `B.to_dual'` at all

Reply:
The only argument i can see for `M₃ →ₗ[R₃] module.dual R₃ M₃` is that you get dot notation on the result to access things in the `module.dual` namespace.

Reply:
> The only argument i can see for `M₃ →ₗ[R₃] module.dual R₃ M₃` is that you get dot notation on the result to access things in the `module.dual` namespace.

But we can still use this, right?  Because we'll still have `to_dual`, just not `to_dual'`.

Reply:
Ah, I missed the `'`